### ISSUE TITLE
Add Meeting Presence

### ISSUE THREAD

**Issue by mattkrick** [ParabolInc/action#103]
## Issue - Enhancement

When a team member joins a meeting, they become a `Participant`. Each `Participant` can have 1 or more `Connection`s, which each have a unique socketId.

This relationship (Meeting:Participant:Connection) could make for a big complicated mess if we have to persist it in the DB. Luckily, I don't think we have to persist anything. Since socketCluster allows clients to publish to channels, we can create our own heartbeat. The logic looks like this:

1. When a team member joins a meeting, they subscribe to a meeting object via Cashay (out of scope)
2. That team member also joins a channel called `presence/:meetingId`
3. socketCluster's `MIDDLEWARE_SUBSCRIBE` checks the JWT auth.
4. After joining, the team member publishes `ROLLCALL` to the channel. 
5. socketCluster's `MIDDLEWARE_PUBLISH_IN` makes sure they have auth to publish & replaces the ROLLCALL event with `{type: 'ROLLCALL', user: authToken.sub}`. That way, we guarantee the authenticity of the user making the request (`authToken.sub === cachedUserId`).
6. `MIDDLEWARE_PUBLISH_OUT` removes the sending socket from the list of receiving sockets. (socket.io calls this a broadcast IIRC).
7. Every 10 seconds (default), the client publishes a `PRESENT` message. When a message with `type === 'ROLLCALL'` is received on the client, the client responds immediately, and resets the interval.
8. `MIDDLEWARE_PUBLISH_IN` again checks auth & replaces `PRESENT` with {type: 'PRESENT', user: authToken.sub}.
9. `MIDDLEWARE_PUBLISH_OUT` removes the sending socket from the list of receiving sockets. 
10. When the server receives an `UNSUBSCRIBE` from `presence/:meetingId`, it publishes `{type: 'LEAVE', user: authToken.sub}. 

The above can be abstracted into its own npm package (@jondubois, any feedback before I publish something like this to npm?)

The following is application logic:
1. Create a `meetingPresence` redux duck. The state will look like: `{[message.user]: Date.now()}`.
2. When a message with `type === 'PRESENT'` comes in, the reducer either updates the `Date.now()`, or adds a new field if `!state.hasOwnProperty(message.user)`.
3. A higher order component that maps this duck to a prop has a `setInterval` in the constructor that takes the state as input and outputs an object where values are an ENUM ('Good', 'Poor', 'Bad'). Those values are calculated by `Date.now() - state[key]`. Good is < 10000. Poor is < 30000. Bad is < 45000. At 60000, the socket will disconnect & that user will be dropped from the list (this is server logic).
4. The new object is passed down through props as `presence`.

- **Estimated effort:** 2 points ([see CONTRIBUTING.md](https://github.com/ParabolInc/action/blob/master/CONTRIBUTING.md#points-and-sizes))
- **(Optional) link to mission spec:** ([duplicate and edit template](https://docs.google.com/document/d/1lgTJnJzBt9-4If883iOvthmc0hy-6C6hBF7KNN-lsXU/edit))

**Comment by mattkrick** [comment]
@jordanh this is next up! Care to poke any holes in it before I get started?

**Comment by jordanh** [comment]
No holes! When I woke up and read it, I loved it!

**Comment by mattkrick** [comment]
Since we don't store the `teamId` in the JWT, we're gonna have to authorize the user somehow. Can't have randos publishing to other people's channels. Assuming we don't want to hit the database for each & every publish, we should probably save it in memory. For now, I'll stick it on `scSocket.authorization`, which in our case will be an array of `teamMember`. I wanna keep the field (`authorization`) vague because this should be a standard pattern that lends itself to other applications. In doing so, we've got a really small JWT which is great for HTTP requests, and we're 1 DB lookup away from getting a whole new world of authorization (ie this is where we'll keep roles & permissions).

**Comment by mattkrick** [comment]
@jordanh if we keep `teamMembers` in memory, we're gonna have a problem keeping it synced up with RethinkDB (the Single Truth&trade;). We've got a couple options:
- use a changefeed, triggered inside the socketCluster middleware. Simple, but a little more memory intensive.
- the first time they hit middleware, stash the whole array in memory. Then, if they don't have authorization, hit the DB again. This is pretty darn nice, except for when someone gets axed from a team _and they're online_, we'd have to find all of their sockets & rip out that membership.

I'm leaning towards the second approach & marking the real-time removal as out of scope for the MVP. thoughts?

**Comment by jordanh** [comment]
The way I always imagined we'd evolve our architecture is to be _real_ naive about where we store objects and store *everything* in RethinkDB and publish *everything* from RethinkDB as a changefeed. Over time this, of course, will suck wind. But it will suck wind with intention.

This of course would have the nice side-effect of keeping everything in sync with our single source of truth. I'm also deadly curious to see how well/poorly rethinkdb scales when you _really_ torture it.

Over time, I always imagined that we'd migrate _all_ of our shared domain-state to a memory-only store like redis or publish messages on a message bus like zeromq.

So, my opinion is do neither, but rather keep in mind we'll port all domain-state to another architecture soonish (likely just following August).

What do you think?

**Comment by mattkrick** [comment]
i am always in favor of a do-it-later approach 👍

**Comment by mattkrick** [comment]
Throwing a wildcard in here & going to the alive-until-proven-dead-approach. Why? Because channeling all of a user's socket connections into 1 channel is going to _very_ advantageous for us in the future, if not immediate future. Everything from analytics to private messages to scaling begins with a way of grouping connections into participants. Also, it lets us go holistic & group on `cachedUserId` instead of `teamMemberId`. Personally, I just like it because we aren't creating duplicate heartbeat logic. For example, if i change my heartbeat timeout from 60 to 45, with this approach I don't have to update the presence logic. Also, it allows us to use cashay & GraphQL. This adds a millisecond or 2 to the server compute time, but the benefit of a schema + keeping domain state inside cashay is too good to pass up.

Here's how it looks on the client:
- When I join a meeting, I subscribe to `meeting/:meetingId` a la cashay, same as above
- I subscribe to `presence/:meetingId` a la cashay. Basically, I want to know when someone makes a rollcall & all I want is a scalar that holds the userId:
```
subscription($meetingId: ID!) {
  rollcall(meetingId: $meetingId) // returns cachedUserId
}
```
- When I get a document from this, I don't persist it. I just reply:
```
mutation($teamId: ID!, $userId: ID!) {
  present(teamId: $teamId, userId: $userId)
}
```
- I subscribe to `participant/:cachedUserId` a la cashay. Basically, I want to know when someone wants to talk to me, regarding which team, on all of my many connections. The unintuitive part here is that I'm going to be subscribing to a channel that accepts multiple teamIds & sending all of them back to each socket connection. For example, I'll send `TeamA` info down to a device that only cares about `TeamB`. Or maybe I can filter it inside the resolve, we'll see.
```
subscription($userId: ID!, $teamId: ID!) {
  participant(userId: $userId) {
    user,
    teamId
  }
}
```

- Now that I'm listening to others who join my meeting & others who want to talk to me, I can ask them who is in the meeting:
```
mutation($meetingId: ID!) {
  soundoff(meetingId: $meetingId)
}
```

**Comment by mattkrick** [comment]
Describing how this works before we close & I forget the gory details:

- When a user joins a meeting, they create a socket connection.
- They subscribe to a channel called `presence/:meetingId`
- The server has some subscription middleware to verify that they can listen on that channel
- When the client gets word that they can listen, it then calls a `SOUNDOFF` in a `cashay.mutate`
- That soundoff contains the user's socket id, called `targetId`. 
- It triggers 2 publishes: 
 - 1 is a `SOUNDOFF` that goes to everyone in the room except (except for `targetId`).
 - The second is a `PRESENT`, which announces to the entire room the `socketId` & `userId` of the person that just joined
- When a client receives `SOUNDOFF`, they respond with `PRESENT` that goes to only the `targetId`
- When a client receives `PRESENT` (`{socketId, userId}`), they insert that into the cashay state
- When the server receives an `unsubscribe` (also triggered on `disconnect`) from a `presence` channel, it publishes a `LEAVE` and the `socketId` of the person that unsubscribed.
- When the client receives `LEAVE`, they remove that presence object from the subscription array.

Doing it this way has a few advantages:
- I can keep everything in cashay, meaning it is still our 1 true source of domain state
- There is only 1 channel to manage, and 3 messages, so complexity is low
- Leave detection originates from the server. Soft and hard exits are handled the same 
- No client receives a useless message, ever

**Comment by jordanh** [comment]
Closed expertly by #114
